src/dev/lupio/lupio_tty.cc

```diff
@@ -0,0 +1,197 @@
+    terminal->regInterfaceCallback([this]() { dataAvailable(); });
+}
+
+void
+LupioTTY::lupioTTYUpdateIRQ()
+{
+    unsigned int irq;
+
+    irq = (writIntrEn && writChar != -1)
+        || (readIntrEn && readChar != -1);
+
+    if (irq) {
+        DPRINTF(LupioTTY, "LupioTTY InterEvent, interrupting\n");
+        platform->postConsoleInt();
+    } else {
+        DPRINTF(LupioTTY, "LupioTTY InterEvent, not interrupting\n");
+        platform->clearConsoleInt();
+    }
+}
+
+void
+LupioTTY::dataAvailable()
+{
+    gem5_assert(terminal->dataAvailable()); 
+    
+    // prevent from overwritting an unread character
+    if (readChar != -1) {
+        fprintf(stderr, "Dropping characters\n");
+    } else {
+        // read data from the terminal
+        readChar = terminal->readData();
+        lupioTTYUpdateIRQ();
+    }
+}
+
+uint64_t
+LupioTTY::lupioTTYRead(uint8_t addr)
+{
+    uint32_t ret = LUPIO_TTY_INVAL;
+
+    switch (addr >> 2) {
+        case LUPIO_TTY_WRIT:
+            DPRINTF(LupioTTY, "Accessing LUPIO_TTY_WRIT\n");
+            if (writChar != -1) {
+                ret = writChar;
+                writChar = -1;
+                lupioTTYUpdateIRQ();
+            }
+            break;
+        case LUPIO_TTY_READ:
+            DPRINTF(LupioTTY, "Accessing LUPIO_TTY_READ\n");
+            if (readChar != -1) {
+                // return new character
+                ret = readChar;
+                readChar = -1;
+                lupioTTYUpdateIRQ();
+            }
+            break;
+        case LUPIO_TTY_CTRL:
+            DPRINTF(LupioTTY, "Accessing LUPIO_TTY_CTRL\n");
+            ret = 0;
+            if (writIntrEn) {
+                ret |= LUPIO_TTY_WBIT;
+            }
+            if (readIntrEn) {
+                ret |= LUPIO_TTY_RBIT;
+            }
+            break;
+        case LUPIO_TTY_STAT:
+            DPRINTF(LupioTTY, "Accessing LUPIO_TTY_STAT\n");
+            // always ready to write
+            ret = LUPIO_TTY_WBIT;
+            // ready to read if unread character available
+            if (readChar != -1) {
+                ret |= LUPIO_TTY_RBIT;
+            }
+            break;
+        default:
+            panic("Unexpected read to the LupioTTY device at address %d!",
+                    addr);
+            break;
+    }
+
+    return ret;
+}
+
+void
+LupioTTY::lupioTTYWrite(uint8_t addr, uint64_t val64)
+{
+    uint32_t val = val64;
+    uint8_t c = val;
+
+    switch (addr >> 2) {
+        case LUPIO_TTY_WRIT:
+            DPRINTF(LupioTTY, "Accessing Write: LUPIO_TTY_WRIT: %d\n", c);
+            // write data to terminal
+            terminal->writeData(c);
+            writChar = c;
+            lupioTTYUpdateIRQ();
+            break;
+        case LUPIO_TTY_CTRL:
+            // set interrupt enable bits
+            DPRINTF(LupioTTY, "Accessing LUPIO_TTY_CTRL\n");
+            writIntrEn = !!(val & LUPIO_TTY_WBIT);
+            readIntrEn = !!(val & LUPIO_TTY_RBIT);
```

Comment:
You shouldn't need the `!!` (unless I'm mistaken). The cast from `uint8_t` to `bool` should implicitly work.

```suggestion
            writIntrEn = val & LUPIO_TTY_WBIT;
            readIntrEn = val & LUPIO_TTY_RBIT;
```

Reply:
It's a pretty common idiom in the Linux kernel, which I think comes from the fact that the kernel doesn't use `bool` very often (but `int` instead). Here you're right, the conversion to bool should work.